Problems/11051/ans_11051.py

```diff
@@ -0,0 +1,43 @@
+# https://www.acmicpc.net/problem/11051
+
+import sys
+
+
+def solve():
+    DIV = 10007
+    input = sys.stdin.readline
+    N, K = map(int, input().split())
+    # 11050 문제와 같이 재귀로 풀면 N이 큰 경우에 대해 재귀 제한에 걸려서 불가능
+    # Memory 제한조건이 256MB이므로, Bottom-up으로 주어진 Range에 대해 조합 테이블을 생성하기 충분
+    # Python은 Overflow가 존재하지 않으므로 Factorial 계산을 해버려도 무방하지만...
+    # nCk = n-1Ck + n-1Ck-1
+    nCr = {}
+    nCr[(1, 0)] = 1
+    nCr[(1, 1)] = 1
+
+    for i in range(1, N + 1):
+        for j in range(i + 1):
+            if i - 1 < 2 * j:
+                j = i - j
+            if j == 1:
+                nCr[(i, j)] = i
+            elif j == 0:
+                nCr[(i, j)] = 1
+            else:
+                # 이부분 머리아파서 정리가 안됨... 걍 빡코딩
+
+                try:
+                    t1 = nCr[(i - 1, j - 1)]
+                except:
+                    t1 = nCr[(i - 1, i - j)]
+                try:
+                    t2 = nCr[(i - 1, j)]
+                except:
+                    t2 = nCr[(i - 1, i - j - 1)]
+                nCr[(i, j)] = (t1 + t2) % DIV
+
+    print(nCr[(N, K)] % DIV)
```

Comment:
아래 code에서 % DIV를 두번하고 있습니다.
실제로 10007이기 때문에 결과에 영향성이 없지만, 
원래 목적대로 라고 한다면 nCr[(i,j)]는 % DIV를 제거하는 것이 맞다고 보여집니다.

```suggestion
                nCr[(i, j)] = (t1 + t2)

    print(nCr[(N, K)] % DIV)
```

Reply:
의도된 설계임. 이항계수의 값은 팩토리얼에 기초하므로 조금만 계수가 커져도 큰 수의 계산을 하게 되는데,
어차피 필요한 것은 10007로 나눈 나머지이고 상기 계산방법에서 더하기만을 하고 있으므로
나머지만 딕셔너리에 적음으로써 데이터 양을 줄이려 한 것.
(a + b) % DIV = (a % DIV + b % DIV) % DIV
상기 식은 a,b,DIV가 양의 정수일 때 항상 성립함.

Reply:
> Comment에 대한 설명 감사합니다. 그런 의도가 있을거라고는 생각하지 못했네요.
>근데 for loop 동안 나누기를 계속 수행하는데 큰 수를 나누더라도 마지막에 1회 하는 것이 runtime 시간이 더 짧지 않을까요? 어떤 계산 방식이나 알고리즘이 코드의 속도에 영향을 주는지 스터디 해주는 것도 좋을거 같네요ㅎㅎㅎ

Overflow에 대해 자유로운 language가 아닌 language보다 적다는걸 생각해봤을 때,
큰 수 계산을 신경 안쓰고 알고리즘을 공부하는건 좋은 어프로치가 아니라는 개인적인 생각에 따라...
파이썬 상에서 효율적인 코드보다는 범용적으로 효율적인 코드를 만드는 어프로치를 하고싶음.